### PR TITLE
Nakadi Auto DLQ

### PR DESCRIPTION
**Pull request by adyach**
# One-line summary
Automatically skip events which can not be processed by consumer. The unprocessable events are identified by number of failed commits.

### CODE REVIEW

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/StreamingContext.java [hunk 1]
```diff
@@ -114,6 +117,11 @@ private StreamingContext(final Builder builder) {
         this.kafkaRecordDeserializer = builder.kafkaRecordDeserializer;
         this.eventTypeCache = builder.eventTypeCache;
         this.featureToggleService = builder.featureToggleService;
+
+        this.userFailedCommitLimit = Optional.ofNullable(getSubscription().getAnnotations())
+          .map(ans -> ans.get(DeadLetterAnnotationValidator.AUTO_DLQ_FAILED_COMMIT_LIMIT))
+          .map(Integer::valueOf)
+          .orElse(null);
```
Comment (a1exsh) [inline api-consumption/src/main/java/org/zalando/nakadi/service/subscription/StreamingContext.java:124]: Whitespace looks out of surrounding style here (expected 8 spaces on continuation).

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/ClosingState.java [hunk 2]
```diff
@@ -62,19 +63,49 @@ public void onEnter() {
...
+
+        getZk().updateTopology(topology -> {
+                    try {
+                        final Partition[] array = Arrays.stream(topology.getPartitions())
+                                .filter(p -> uncommittedOffsets.containsKey(
+                                        new EventTypePartition(p.getEventType(), p.getPartition())))
+                                .map(Partition::toPartitionWithIncFailedCommits)
+                                .toArray(Partition[]::new);
+                        return array;
+                    } catch (RuntimeException re) {
```
Comment (a1exsh) [inline api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/ClosingState.java:101]: We don't really expect a runtime exception here, do we?

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/PartitionData.java [hunk 3]
```diff
@@ -53,7 +53,8 @@ class PartitionData {
 
     @Nullable
     List<ConsumedEvent> takeEventsToStream(final long currentTimeMillis,
-                                           final int batchSize, final long batchTimeoutMillis,
+                                           final int batchSize,
```
Comment (a1exsh) [inline api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/PartitionData.java:56]: unrelated change, let's revert it

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StartingState.java [hunk 4]
```diff
@@ -14,6 +16,9 @@
 import java.util.stream.Collectors;
 
 public class StartingState extends State {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StartingState.class);
```
Comment (a1exsh) [inline api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StartingState.java:20]: not used

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java [hunk 5]
```diff
@@ -232,11 +234,20 @@ private void rememberEvent(final ConsumedEvent event) {
...
+            return 1;
+        }
+
         final long unconfirmed = offsets.values().stream().mapToLong(PartitionData::getUnconfirmed).sum();
-        final long limit = getParameters().maxUncommittedMessages - unconfirmed;
+        final long limit = getMaxUncommittedMessages() - unconfirmed;
         return getParameters().getMessagesAllowedToSend(limit, this.sentEvents);
     }
 
+    private int getMaxUncommittedMessages() {
```
Comment (a1exsh) [inline api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java:247]: only used once?

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java [hunk 6]
```diff
@@ -261,15 +272,64 @@ private void streamToOutput(final boolean streamTimeoutReached) {
...
+
+            while (true) {
+                if (partition != null && partition.isLookingForDeadLetter()) {
+                    final NakadiCursor lastDeadLetterCursor = getContext().getCursorConverter().convert(
+                            partition.getEventType(),
+                            new Cursor(partition.getPartition(), partition.getLastDeadLetterOffset()));
+                    if (getComparator().compare(partitionData.getCommitOffset(), lastDeadLetterCursor) >= 0) {
+                        getZk().updateTopology(topology -> Arrays.stream(topology.getPartitions())
+                                .filter(p -> p.getPartition().equals(etp.getPartition()))
+                                .map(p -> p.toLookingDeadLetter(null))
```
Comment (a1exsh) [inline api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java:287]: let's rename it accordingly: `toLastDeadLetterOffset`?

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java [hunk 7]
```diff
@@ -314,6 +374,10 @@ private void streamToOutput(final boolean streamTimeoutReached) {
         }
     }
 
+    private int getBatchLimitEvents() {
```
Comment (a1exsh) [inline api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java:377]: probably also used only once

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java [hunk 8]
```diff
@@ -324,7 +388,22 @@ private Optional<String> getDebugMessage(final Map.Entry<EventTypePartition, Par
...
+            final String failedCommitsPartitions = Arrays.stream(getZk().getTopology().getPartitions())
+                    .filter(p -> p.getFailedCommitsCount() > 0 || p.isLookingForDeadLetter())
+                    .map(Partition::toString)
+                    .collect(Collectors.joining(" "));
+
+            if (failedCommitsPartitions != null && !failedCommitsPartitions.isEmpty()) {
+                sb.append(". Failed commits: ").append(failedCommitsPartitions);
+                LOG.error(failedCommitsPartitions);
+            } else {
+                LOG.error("No failed commits");
```
Comment (a1exsh) [inline api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java:403]: let's remove these logs (or demote them to `debug` level)

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java [hunk 9]
```diff
@@ -663,15 +764,25 @@ private void reassignCommitted() {
         }
     }
 
-    void offsetChanged(final EventTypePartition key) {
-        if (offsets.containsKey(key)) {
-            final PartitionData data = offsets.get(key);
+    void offsetChanged(final EventTypePartition etp) {
```
Comment (a1exsh) [inline api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java:767]: (unrelated) wonder why it's not `private`?

File: app/src/main/resources/kpi_event_types.json [hunk 10]
```diff
@@ -82,8 +82,8 @@
     "default_statistic": {
       "messages_per_minute": 100,
       "message_size": 100,
-      "read_parallelism": 8,
-      "write_parallelism": 8
+      "read_parallelism": 1,
+      "write_parallelism": 1
```
Comment (a1exsh) [inline app/src/main/resources/kpi_event_types.json:86]: revert?

File: core-common/src/main/java/org/zalando/nakadi/annotations/validation/DeadLetterAnnotationValidator.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
+package org.zalando.nakadi.annotations.validation;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import java.util.Map;
+
+public class DeadLetterAnnotationValidator implements
+        ConstraintValidator<DeadLetterValidAnnotations, Map<String, String>> {
+
+    public static final String AUTO_DLQ_FAILED_COMMIT_LIMIT = "nakadi.zalando.org/auto-dlq-failed-commit-limit";
```
Comment (a1exsh) [inline core-common/src/main/java/org/zalando/nakadi/annotations/validation/DeadLetterAnnotationValidator.java:10]: The name `auto-dlq` is not entirely correct, as we have only implemented skipping the events silently.
Should we change it to `auto-skip-event-failed-commit-limit` and make `dlq` a choice in another annotation value, e.g. `auto-skip-event-policy`?

File: core-common/src/main/java/org/zalando/nakadi/service/subscription/model/Partition.java [hunk 12]
```diff
@@ -35,6 +34,11 @@ public String getDescription() {
     private String nextSession;
     @JsonProperty("state")
     private State state;
+    // todo create data struct
```
Comment (a1exsh) [inline core-common/src/main/java/org/zalando/nakadi/service/subscription/model/Partition.java:37]: still need that todo?

File: core-common/src/main/java/org/zalando/nakadi/service/subscription/model/Partition.java [hunk 13]
```diff
@@ -52,8 +56,47 @@ public Partition(
...
+    }
+
+    public Partition toLookingDeadLetter(final String lastDeadLetterOffset) {
+
+        if (lastDeadLetterOffset != null && this.lastDeadLetterOffset == null) {
+            // failed commits reset to count for failures of the specific event
+            return new Partition(eventType, partition, session, nextSession, state, 0, lastDeadLetterOffset);
+        } else if (lastDeadLetterOffset == null && this.lastDeadLetterOffset != null) {
+            return new Partition(eventType, partition, session, nextSession, state, 0, null);
+        }
```
Comment (a1exsh) [inline core-common/src/main/java/org/zalando/nakadi/service/subscription/model/Partition.java:97]:
```suggestion
        if ((lastDeadLetterOffset != null && this.lastDeadLetterOffset == null) || (lastDeadLetterOffset == null && this.lastDeadLetterOffset != null)) {
            return new Partition(eventType, partition, session, nextSession, state, 0, lastDeadLetterOffset);
        }
```
In both cases we reset failed commit count to 0, and use the given value of `lastDeadLetterOffset` (either it is `null` or not).

File: core-common/src/main/java/org/zalando/nakadi/service/subscription/model/Partition.java [hunk 14]
```diff
@@ -134,7 +177,8 @@ public String getEffectiveSession() {
 
     @Override
     public String toString() {
-        return eventType + ":" + partition + "->" + state + ":" + session + "->" + nextSession;
+        return eventType + ":" + partition + "->" + state + ":" +
+                session + "->" + nextSession + ":" + failedCommitsCount + ":" + lastDeadLetterOffset;
```
Comment (a1exsh) [inline core-common/src/main/java/org/zalando/nakadi/service/subscription/model/Partition.java:181]:
```suggestion
        return eventType + ":" + partition + "->" + state + ":" +
                session + "->" + nextSession + "->" + failedCommitsCount + ":" + lastDeadLetterOffset;
```
It's still cryptic, but more consistent this way ;)

File: core-common/src/test/java/org/zalando/nakadi/service/subscription/model/PartitionTest.java [hunk 15]
```diff
@@ -57,10 +57,10 @@ public void moveReassigningPartitionShouldPutToAssignedState() {
         final String partition = "partition";
 
         ImmutableList.of(
-                new Partition(eventType, partition, "x", "x1", Partition.State.REASSIGNING),
-                new Partition(eventType, partition, "x", "T", Partition.State.REASSIGNING),
-                new Partition(eventType, partition, "T", "x", Partition.State.REASSIGNING),
-                new Partition(eventType, partition, "T", "T1", Partition.State.REASSIGNING))
+                        new Partition(eventType, partition, "x", "x1", Partition.State.REASSIGNING),
```
Comment (a1exsh) [inline core-common/src/test/java/org/zalando/nakadi/service/subscription/model/PartitionTest.java:60]: unrelated

**Review by a1exsh — COMMENTED** [review]
Please check the comments :)

File: api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java (hunk 9 above)
Comment (adyach) [reply api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java:767]: used in test in the same package

**Comment by a1exsh** [comment]
:+1:

**Comment by adyach** [comment]
👍
